File: src/__tests__/SendQueue.test.ts
// WebSocketRT: Real-time websocket library
// Copyright (c) Leo C. Singleton IV <leo@leosingleton.com>
// See LICENSE in the project root for license information.

import { Message } from './Message';
import { SendQueue } from './SendQueue';
import { OutgoingMessage } from './OutgoingMessage';

describe('SendQueue', () => {

  it('Ensures the SendQueue returns messages in priority order', () => {
    const queue = new SendQueue(16);
    const message1 = new OutgoingMessage(1, new Message(100), 4);
    const message2 = new OutgoingMessage(2, new Message(100), 2); // 2 is higher-priority than 4

    queue.enqueue(message1);
    queue.enqueue(message2);

    // message2 should be returned first, as it is higher priority
    const result1 = queue.getNext(100);
    expect(result1.message).toEqual(message2);
    expect(result1.bytesToSend).toEqual(100);
    result1.message._bytesSent += result1.bytesToSend;

    // message1 should be returned next
    const result2 = queue.getNext(100);
    expect(result2.message).toEqual(message1);
    expect(result2.bytesToSend).toEqual(100);
    result2.message._bytesSent += result2.bytesToSend;

    // null should be returned, as there are no more messages
    const result3 = queue.getNext(100);
    expect(result3.message).toBeNull();
    expect(result3.bytesToSend).toEqual(0);
  });

  it("Tests the SendQueue's Cancel() method", () => {
    const queue = new SendQueue(16);

    // Create some messages
    let messageNumber = 0;
    const priorities = [ 1, 2, 3, 3, 3, 15 ];
    const messages: OutgoingMessage[] = [];
    for (const priority of priorities) {
      const message = new OutgoingMessage(messageNumber, new Message(100), priority);
      queue.enqueue(message);
      messages[messageNumber++] = message;
    }

    // Cancel message 0. Message 1 will be returned next.
    queue.cancel(messages[0]);
    const next = queue.getNext(100);
    expect(next.message.messageNumber).toEqual(1);

    // Cancel message 2. Message 3 will be returned next.
    queue.cancel(messages[2]);
    const next2 = queue.getNext(100);
    expect(next2.message.messageNumber).toEqual(3);

    // Cancel message 5. Message 4 will be returned next.
    queue.cancel(messages[5]);
    const next3 = queue.getNext(100);
    expect(next3.message.messageNumber).toEqual(4);

    // The queue is now empty.
    const next4 = queue.getNext(100);
    expect(next4.message).toBeNull();
  });

  // This repros a bug from Oct 2018 where enqueue() was not updating the _highestPriority value
  it('Sends higher priority messages before lower', () => {
    const queue = new SendQueue(16);

    const message1 = new OutgoingMessage(0, new Message(100), 0); // Higher-priority message
    const message2 = new OutgoingMessage(1, new Message(200), 1); // Lower-priority message

    // Send the lower-priority, and read only half of it
    queue.enqueue(message2);
    const next = queue.getNext(100);
    expect(next.message.messageNumber).toEqual(1);
    expect(next.bytesToSend).toEqual(100);
    next.message._bytesSent += next.bytesToSend;

    // Send the higher-priority. It should preempt the lower-priority.
    queue.enqueue(message1);
    const next2 = queue.getNext(100);
    expect(next2.message.messageNumber).toEqual(0);
    expect(next2.bytesToSend).toEqual(100);
    next2.message._bytesSent += next2.bytesToSend;

    // Now we should get the remainder of the lower-priority.
    const next3 = queue.getNext(100);
    expect(next3.message.messageNumber).toEqual(1);
    expect(next3.bytesToSend).toEqual(100);
    next3.message._bytesSent += next3.bytesToSend;

    // The queue is now empty.
    const next4 = queue.getNext(100);
    expect(next4.message).toBeNull();
    expect(next4.bytesToSend).toEqual(0);
  });

});

File: dotnet/UnitTests/SendQueueTest.cs
using Xunit;

namespace LeoSingleton.WebSocketRT.UnitTests
{
    public class SendQueueTest
    {
        /// <summary>
        /// Ensures the SendQuee returns messages in priority order
        /// </summary>
        [Fact]
        public void CheckPriority()
        {
            var queue = new SendQueue(16);
            var message1 = new OutgoingMessage(1, new Message(100), 4);
            var message2 = new OutgoingMessage(2, new Message(100), 2); // 2 is higher-priority than 4

            queue.Enqueue(message1);
            queue.Enqueue(message2);

            OutgoingMessage message;
            int result;

            // message2 should be returned first, as it is higher priority
            result = queue.GetNext(100, out message);
            Assert.Equal(message2, message);
            Assert.Equal(100, result);
            message._BytesSent += result;

            // message1 should be returned next
            result = queue.GetNext(100, out message);
            Assert.Equal(message1, message);
            Assert.Equal(100, result);
            message._BytesSent += result;

            // null should be returned, as there are not more messages
            result = queue.GetNext(100, out message);
            Assert.Null(message);
            Assert.Equal(0, result);
        }

        /// <summary>
        /// Tests the SendQueue's Cancel() method
        /// </summary>
        [Fact]
        public void TestCancellation()
        {
            var queue = new SendQueue(16);

            // Create some messages
            byte messageNumber = 0;
            var priorities = new int[6] { 1, 2, 3, 3, 3, 15 };
            var messages = new OutgoingMessage[6];
            foreach (int priority in priorities)
            {
                var message = new OutgoingMessage(messageNumber, new Message(100), priority);
                queue.Enqueue(message);
                messages[messageNumber++] = message;
            }

            OutgoingMessage msg;

            // Cancel message 0. Message 1 will be returned next.
            queue.Cancel(messages[0]);
            queue.GetNext(100, out msg);
            Assert.Equal(1, msg.MessageNumber);

            // Cancel message 2. Message 3 will be returned next.
            queue.Cancel(messages[2]);
            queue.GetNext(100, out msg);
            Assert.Equal(3, msg.MessageNumber);

            // Cancel message 5. Message 4 will be returned next.
            queue.Cancel(messages[5]);
            queue.GetNext(100, out msg);
            Assert.Equal(4, msg.MessageNumber);

            // The queue is now empty.
            queue.GetNext(100, out msg);
            Assert.Null(msg);
        }

        /// <summary>
        /// Ensures higher priority messages are sent before lower
        /// </summary>
        /// <remarks>
        /// This repros a bug from Oct 2018 where Enqueue() was not updating the HighestPriority value. Although the
        /// bug occurred in the TypeScript port, the same equally could have happened in C#.
        /// </remarks>
        [Fact]
        public void TestHigherBeforeLower()
        {
            var queue = new SendQueue(16);

            var message1 = new OutgoingMessage(0, new Message(100), 0); // Higher-priority message
            var message2 = new OutgoingMessage(1, new Message(200), 1); // Lower-priority message

            OutgoingMessage message;
            int result;

            // Send the lower-priority, and read only half of it
            queue.Enqueue(message2);
            result = queue.GetNext(100, out message);
            Assert.Equal(1, message.MessageNumber);
            Assert.Equal(100, result);
            message._BytesSent += result;

            // Send the higher-priority. It should preempt the lower-priority.
            queue.Enqueue(message1);
            result = queue.GetNext(100, out message);
            Assert.Equal(0, message.MessageNumber);
            Assert.Equal(100, result);
            message._BytesSent += result;

            // Now we should get the remainder of the lower-priority.
            result = queue.GetNext(100, out message);
            Assert.Equal(1, message.MessageNumber);
            Assert.Equal(100, result);
            message._BytesSent += result;

            // The queue is now empty.
            result = queue.GetNext(100, out message);
            Assert.Null(message);
            Assert.Equal(0, result);
        }
    }
}
